File: LC/[0684] Redundant Connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        parent = {}

        def find(x):
            if x != parent[x]:
                parent[x] = find(parent[x])
            return parent[x]

        def union(a, b):
            pa = find(a)
            pb = find(b)

            if pa == pb:
                return False

            if pa < pb:
                parent[pb] = pa
            else:
                parent[pa] = pb

            return True

        for a, b in edges:
            if a not in parent:
                parent[a] = a
            if b not in parent:
                parent[b] = b

            if not union(a, b):
                return [a, b]

        return []

File: LC/[0684] Redundant Connection.java
class Solution {
    private Map<Integer, Integer> parent = new HashMap<>();

    public int[] findRedundantConnection(int[][] edges) {
        for (int[] edge : edges) {
            int a = edge[0];
            int b = edge[1];

            if (!parent.containsKey(a)) parent.put(a, a);
            if (!parent.containsKey(b)) parent.put(b, b);

            if (find(a) == find(b)) {
                return edge;
            }

            union(a, b);
        }

        return new int[0];
    }

    private int find(int x) {
        if (parent.get(x) != x) {
            parent.put(x, find(parent.get(x)));
        }
        return parent.get(x);
    }

    private void union(int x, int y) {
        x = find(x);
        y = find(y);

        if (x == y) return;

        if (x < y) {
            parent.put(y, x);
        } else {
            parent.put(x, y);
        }
    }
}